### PR TITLE
Unify finish trajectory logic

### PR DESCRIPTION
**Pull request by jihoonl**
Port active trajectory checks and subscriber shutdown logic into FinishTrajectory(), so that a trajectory can be cleanly finished by function call.

### CODE REVIEW

**Comment by SirVer** [comment]
@jihoonl I generally like the change, however the code contained a bug: The mutex was no longer taken for `Node::FinishTrajectory`, though that is a public method that totally must take it. 

I suggested a refactor. Could you have a look if you still like the code?

File: cartographer_ros/cartographer_ros/node.cc [hunk 1]
```diff
@@ -421,27 +444,7 @@ bool Node::HandleFinishTrajectory(
...
-    return false;
-  }
-
-  // Shutdown the subscribers of this trajectory.
-  for (auto& entry : subscribers_[trajectory_id]) {
-    entry.subscriber.shutdown();
-    subscribed_topics_.erase(entry.topic);
-    LOG(INFO) << "Shutdown the subscriber of [" << entry.topic << "]";
-  }
-  CHECK_EQ(subscribers_.erase(trajectory_id), 1);
```
Comment (SirVer) [inline cartographer_ros/cartographer_ros/node.cc]: This has gone? Should it not be retained?

Comment (jihoonl) [reply cartographer_ros/cartographer_ros/node.cc]: I think there was a reason which I don't remember now. I will retest it when I return to the office.

Comment (jihoonl) [reply cartographer_ros/cartographer_ros/node.cc]: Nevermind. it was by mistake.

**Comment by jihoonl** [comment]
Thanks @SirVer, I like the change and have confirmed that it is working.

**Comment by SirVer** [comment]
Thank you and sorry that it took so long.

@wally-the-cartographer merge

**Comment by wally-the-cartographer** [comment]
Merge requested by authorized user SirVer. Merge queue now has a length of 1.

**Comment by wally-the-cartographer** [comment]
Error: exit status 128

Log:
~~~
=> Rebasing PR 552 on googlecartographer/cartographer_ros onto master.
=> Switching to branch master.
 $ git rebase --abort [data/cartographer_ros]
No rebase in progress?
 $ git checkout -f [data/cartographer_ros]
Your branch is up-to-date with 'origin/master'.
 $ git clean -fd [data/cartographer_ros]
 $ git checkout master [data/cartographer_ros]
Already on 'master'
Your branch is up-to-date with 'origin/master'.
 $ git pull [data/cartographer_ros]
Already up-to-date.
 $ git remote add magazino git@github.com:magazino/cartographer_ros.git [data/cartographer_ros]
 $ git fetch magazino [data/cartographer_ros]
Warning: Permanently added 'github.com,192.30.253.112' (RSA) to the list of known hosts.
From github.com:magazino/cartographer_ros
 * [new branch]      cmakelist_fixes    -> magazino/cmakelist_fixes
 * [new branch]      constraints_visualization -> magazino/constraints_visualization
 * [new branch]      finish_trajectory  -> magazino/finish_trajectory
 * [new branch]      initialpose_topic  -> magazino/initialpose_topic
 * [new branch]      master             -> magazino/master
 * [new branch]      occ_exporter       -> magazino/occ_exporter
 * [new branch]      occpancy_exporter  -> magazino/occpancy_exporter
 * [new branch]      resolve_topic_name -> magazino/resolve_topic_name
 * [new branch]      submap_id_text     -> magazino/submap_id_text
 $ git branch --track pr_552 magazino/finish_trajectory [data/cartographer_ros]
Branch pr_552 set up to track remote branch finish_trajectory from magazino.
=> Switching to branch pr_552.
 $ git rebase --abort [data/cartographer_ros]
No rebase in progress?
 $ git checkout -f [data/cartographer_ros]
Your branch is up-to-date with 'origin/master'.
 $ git clean -fd [data/cartographer_ros]
 $ git checkout pr_552 [data/cartographer_ros]
Switched to branch 'pr_552'
Your branch is up-to-date with 'magazino/finish_trajectory'.
 $ git pull --force [data/cartographer_ros]
Already up-to-date.
 $ git rebase master [data/cartographer_ros]
First, rewinding head to replay your work on top of it...

*** Please tell me who you are.

Run

  git config --global user.email "you@example.com"
  git config --global user.name "Your Name"

to set your account's default identity.
Omit --global to set the identity only in this repository.

fatal: unable to auto-detect email address (got 'root@d3c4dbc8495b.(none)')
=> Abandoning branch pr_552.
=> Switching to branch master.
 $ git rebase --abort [data/cartographer_ros]
No rebase in progress?
 $ git checkout -f [data/cartographer_ros]
 $ git clean -fd [data/cartographer_ros]
 $ git checkout master [data/cartographer_ros]
Switched to branch 'master'
Your branch is up-to-date with 'origin/master'.
 $ git branch -D pr_552 [data/cartographer_ros]
Deleted branch pr_552 (was 00b4ef3).

~~~

**Comment by SirVer** [comment]
@wally-the-cartographer merge

**Comment by wally-the-cartographer** [comment]
Merge requested by authorized user SirVer. Merge queue now has a length of 1.

**Comment by wally-the-cartographer** [comment]
Refusing to merge, since the combined checks state is failure

**Comment by SirVer** [comment]
@jihoonl sorry for the noise. The PR is guinea pig for merge bot deployment.

@wally-the-cartographer

**Comment by jihoonl** [comment]
No problem. Feel free to test

**Comment by wally-the-cartographer** [comment]
Merge requested by authorized user SirVer. Merge queue now has a length of 1.

**Pull request merged**